src/addons/messagelog/src/main/java/ee/ria/xroad/proxy/messagelog/LogManager.java

```diff
@@ -212,62 +202,81 @@ protected Props getTimestamperImpl() {
     }
 
     protected Props getArchiverImpl() {
-        return Props.create(
-            LogArchiver.class,
-            Paths.get(MessageLogProperties.getArchivePath()),
-            Paths.get(SystemProperties.getTempFilesPath())
-        );
+        return Props.create(LogArchiver.class, Paths.get(MessageLogProperties.getArchivePath()),
+                Paths.get(SystemProperties.getTempFilesPath()));
     }
 
     protected Props getCleanerImpl() {
         return Props.create(LogCleaner.class);
     }
 
-    protected TimestampRecord timestampImmediately(MessageRecord logRecord)
-            throws Exception {
-        log.trace("timestampImmediately({})", logRecord);
+    protected Props getTimestampingPostWorkerImpl() {
+        return Props.create(TimestampingPostWorker.class, this).withDispatcher("akka.control-aware-dispatcher");
+    }
 
-        try {
+    /**
+     * Timestamping post worker handling SetTimestampingStatusMessage and SaveTimestampedDataMessage messages.
+     */
+    @RequiredArgsConstructor
+    public static class TimestampingPostWorker extends UntypedActor {
+
+        private final LogManager logManager;
+
+        @Override
+        public void onReceive(Object message) throws Exception {
+            log.trace("TimestampingPostWorker::onReceive({})", message);
 
+            try {
+                if (message instanceof SetTimestampingStatusMessage) {
+                    logManager.setTimestampingStatus((SetTimestampingStatusMessage) message);
```

Comment:
An actor should not directly call a method of an another actor (it should send a message). In addition, this is a potential concurrency issue, since LogManager's internal state is modified.

Reply:
For the sake of clarity and comprehensibility the communication between LogManager and TaskQueue is completely refactored now and also some additional bugs (related with timestamp-immediately mode) are fixed. It seems, that initial performance issue has been arised from pull request finish-6.7.10, where previous solution was rebuild and timestamping post processing (modification of DB records) was moved from TaskQueue actor to the LogManager actor.